src/core/utilities/cellmap.hpp

```diff
@@ -0,0 +1,565 @@
+    // with at its right cell.
+    // It could not be in the map at 'oldCell', in that case we want to add it.
+    auto& oldList = cellIndexes_(local_(oldCell));
+    if (oldList.is_indexed(itemIndex))
+        oldList.remove(itemIndex);
+    add(items, itemIndex);
+}
+
+
+
+
+template<std::size_t dim, typename cell_index_t>
+template<typename Range, typename Predicate, typename CellExtractor>
+inline auto CellMap<dim, cell_index_t>::partition(Range range, Predicate&& pred,
+                                                  CellExtractor extract)
+{
+    std::size_t toSwapIndex = range.iend() - 1;
+    auto pivot              = range.iend();
+    for (auto const& cell : box_)
+    {
+        auto& itemIndexes = cellIndexes_(local_(cell));
+        if (!pred(cell))
+        {
+            for (auto currentIdx : itemIndexes)
+            {
+                // partition only indexes in range
+                if (currentIdx >= range.ibegin() and currentIdx < range.iend())
+                {
+                    assert(pivot > 0);
+                    --pivot;
+                    if (currentIdx < toSwapIndex)
+                    {
+                        while (!pred(extract(range.array()[toSwapIndex]))
+                               and toSwapIndex > currentIdx)
+                        {
+                            --toSwapIndex;
+                        }
+                        if (toSwapIndex > currentIdx)
+                        {
+                            assert(pred(extract(range.array()[toSwapIndex])));
+                            assert(toSwapIndex >= range.ibegin());
+                            itemIndexes.updateIndex(currentIdx, toSwapIndex);
+                            auto& l = cellIndexes_(local_(extract(range.array()[toSwapIndex])));
+                            l.updateIndex(toSwapIndex, currentIdx);
+                            std::swap(range.array()[currentIdx], range.array()[toSwapIndex]);
+                            --toSwapIndex;
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    return makeRange(range.array(), range.ibegin(), range.ibegin() + pivot);
+}
+
+
+
+
+template<std::size_t dim, typename cell_index_t>
+template<typename Range>
+inline void CellMap<dim, cell_index_t>::erase(Range&& range)
+{
+    auto& items = range.array();
+
+    // first erase indexes from the cellmap
+    // then items from the array
+    for (std::size_t i = range.ibegin(); i < range.iend(); ++i)
+    {
+        erase(items, i);
+    }
+    items.erase(range.begin(), range.end());
+}
+
+
+
+
+template<std::size_t dim, typename cell_index_t>
+template<typename Array, typename CellIndex, typename CellExtractor>
+inline void CellMap<dim, cell_index_t>::erase(CellIndex cell, Array& items, CellExtractor extract)
+{
+    auto lastIndex     = items.size() - 1;
+    auto& itemsToErase = cellIndexes_(local_(cell));
+    assert(false);
+    //  for (auto itemIndex : itemsToErase)
```

Comment:
is this function supposed to exist?

Reply:
I cautiously answer :  no and it should not...
reason is that this function asserts false, cellmap is only visible through the particleArray...
 the other overload `     auto erase(IndexRange_&& range)` in ParticleArray calls `inline void CellMap<dim, cell_index_t>::erase(Range&& range)`, which seems to remove all indexes in the range from the map, and all corresponding items from the container associated with the range....

removed... :crossed_fingers: